File: examples/go/math-demo/MathDemo_test.go
package contract

import (
	"testing"
	"tsop"
)

func TestMathDemo_Safediv(t *testing.T) {
	c := &MathDemo{Value: 100}
	c.DivideBy(5)
	if c.Value != 20 {
		t.Errorf("expected 20, got %d", c.Value)
	}
}

func TestMathDemo_Safediv_Truncates(t *testing.T) {
	c := &MathDemo{Value: 7}
	c.DivideBy(2)
	if c.Value != 3 {
		t.Errorf("expected 3, got %d", c.Value)
	}
}

func TestMathDemo_Safediv_RejectsDivByZero(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected panic on division by zero")
		}
	}()
	c := &MathDemo{Value: 10}
	c.DivideBy(0)
}

func TestMathDemo_PercentOf(t *testing.T) {
	c := &MathDemo{Value: 10000}
	c.WithdrawWithFee(1000, 500) // 5% fee = 50, total = 1050
	if c.Value != 8950 {
		t.Errorf("expected 8950, got %d", c.Value)
	}
}

func TestMathDemo_Clamp_Below(t *testing.T) {
	c := &MathDemo{Value: 3}
	c.ClampValue(10, 100)
	if c.Value != 10 {
		t.Errorf("expected 10, got %d", c.Value)
	}
}

func TestMathDemo_Clamp_Above(t *testing.T) {
	c := &MathDemo{Value: 200}
	c.ClampValue(10, 100)
	if c.Value != 100 {
		t.Errorf("expected 100, got %d", c.Value)
	}
}

func TestMathDemo_Clamp_InRange(t *testing.T) {
	c := &MathDemo{Value: 50}
	c.ClampValue(10, 100)
	if c.Value != 50 {
		t.Errorf("expected 50, got %d", c.Value)
	}
}

func TestMathDemo_Sign_Positive(t *testing.T) {
	c := &MathDemo{Value: 42}
	c.Normalize()
	if c.Value != 1 {
		t.Errorf("expected 1, got %d", c.Value)
	}
}

func TestMathDemo_Sign_Negative(t *testing.T) {
	c := &MathDemo{Value: -7}
	c.Normalize()
	if c.Value != -1 {
		t.Errorf("expected -1, got %d", c.Value)
	}
}

func TestMathDemo_Sign_Zero(t *testing.T) {
	c := &MathDemo{Value: 0}
	c.Normalize()
	if c.Value != 0 {
		t.Errorf("expected 0, got %d", c.Value)
	}
}

func TestMathDemo_Pow(t *testing.T) {
	c := &MathDemo{Value: 2}
	c.Exponentiate(10)
	if c.Value != 1024 {
		t.Errorf("expected 1024, got %d", c.Value)
	}
}

func TestMathDemo_Pow_Zero(t *testing.T) {
	c := &MathDemo{Value: 99}
	c.Exponentiate(0)
	if c.Value != 1 {
		t.Errorf("expected 1, got %d", c.Value)
	}
}

func TestMathDemo_Sqrt(t *testing.T) {
	c := &MathDemo{Value: 100}
	c.SquareRoot()
	if c.Value != 10 {
		t.Errorf("expected 10, got %d", c.Value)
	}
}

func TestMathDemo_Sqrt_NonPerfect(t *testing.T) {
	c := &MathDemo{Value: 10}
	c.SquareRoot()
	if c.Value != 3 {
		t.Errorf("expected 3, got %d", c.Value)
	}
}

func TestMathDemo_Gcd(t *testing.T) {
	c := &MathDemo{Value: 12}
	c.ReduceGcd(8)
	if c.Value != 4 {
		t.Errorf("expected 4, got %d", c.Value)
	}
}

func TestMathDemo_Gcd_Coprime(t *testing.T) {
	c := &MathDemo{Value: 7}
	c.ReduceGcd(13)
	if c.Value != 1 {
		t.Errorf("expected 1, got %d", c.Value)
	}
}

func TestMathDemo_MulDiv(t *testing.T) {
	c := &MathDemo{Value: 1000}
	c.ScaleByRatio(3, 4)
	if c.Value != 750 {
		t.Errorf("expected 750, got %d", c.Value)
	}
}

func TestMathDemo_Log2(t *testing.T) {
	c := &MathDemo{Value: 1024}
	c.ComputeLog2()
	if c.Value != 10 {
		t.Errorf("expected 10, got %d", c.Value)
	}
}

func TestMathDemo_Compile(t *testing.T) {
	if err := tsop.CompileCheck("MathDemo.tsop.go"); err != nil {
		t.Fatalf("TSOP compile check failed: %v", err)
	}
}

File: examples/rust/math-demo/MathDemo_test.rs
#[path = "MathDemo.tsop.rs"]
mod contract;

use contract::*;

#[test]
fn test_safediv() {
    let mut m = MathDemo { value: 100 };
    m.divide_by(5);
    assert_eq!(m.value, 20);
}

#[test]
fn test_safediv_truncates() {
    let mut m = MathDemo { value: 7 };
    m.divide_by(2);
    assert_eq!(m.value, 3);
}

#[test]
#[should_panic(expected = "division by zero")]
fn test_safediv_rejects_zero() {
    let mut m = MathDemo { value: 10 };
    m.divide_by(0);
}

#[test]
fn test_percent_of() {
    let mut m = MathDemo { value: 10000 };
    m.withdraw_with_fee(1000, 500); // 5% fee = 50, total = 1050
    assert_eq!(m.value, 8950);
}

#[test]
fn test_clamp_below() {
    let mut m = MathDemo { value: 3 };
    m.clamp_value(10, 100);
    assert_eq!(m.value, 10);
}

#[test]
fn test_clamp_above() {
    let mut m = MathDemo { value: 200 };
    m.clamp_value(10, 100);
    assert_eq!(m.value, 100);
}

#[test]
fn test_clamp_in_range() {
    let mut m = MathDemo { value: 50 };
    m.clamp_value(10, 100);
    assert_eq!(m.value, 50);
}

#[test]
fn test_sign_positive() {
    let mut m = MathDemo { value: 42 };
    m.normalize();
    assert_eq!(m.value, 1);
}

#[test]
fn test_sign_negative() {
    let mut m = MathDemo { value: -7 };
    m.normalize();
    assert_eq!(m.value, -1);
}

#[test]
fn test_sign_zero() {
    let mut m = MathDemo { value: 0 };
    m.normalize();
    assert_eq!(m.value, 0);
}

#[test]
fn test_pow() {
    let mut m = MathDemo { value: 2 };
    m.exponentiate(10);
    assert_eq!(m.value, 1024);
}

#[test]
fn test_pow_zero() {
    let mut m = MathDemo { value: 99 };
    m.exponentiate(0);
    assert_eq!(m.value, 1);
}

#[test]
fn test_sqrt() {
    let mut m = MathDemo { value: 100 };
    m.square_root();
    assert_eq!(m.value, 10);
}

#[test]
fn test_sqrt_non_perfect() {
    let mut m = MathDemo { value: 10 };
    m.square_root();
    assert_eq!(m.value, 3);
}

#[test]
fn test_gcd() {
    let mut m = MathDemo { value: 12 };
    m.reduce_gcd(8);
    assert_eq!(m.value, 4);
}

#[test]
fn test_gcd_coprime() {
    let mut m = MathDemo { value: 7 };
    m.reduce_gcd(13);
    assert_eq!(m.value, 1);
}

#[test]
fn test_mul_div() {
    let mut m = MathDemo { value: 1000 };
    m.scale_by_ratio(3, 4);
    assert_eq!(m.value, 750);
}

#[test]
fn test_log2() {
    let mut m = MathDemo { value: 1024 };
    m.compute_log2();
    assert_eq!(m.value, 10);
}

#[test]
fn test_compile() {
    tsop::compile_check(include_str!("MathDemo.tsop.rs"), "MathDemo.tsop.rs").unwrap();
}
